chore(ci): enforce shellscript formatting and style

jamilbk opened the pull request:
Noticed that we all have different styles of writing scripts :-).

This PR adds linting to our shell scripts to standardize on formatting, catch common issues and/or possible security bugs.

For editor setup:
- Ensure [`shellcheck`](https://github.com/koalaman/shellcheck) and [`shfmt`](https://github.com/mvdan/sh) are in your `PATH`
- Configure `shfmt` with indentation of `4`, otherwise it uses tabs by default. [Here](https://github.com/jamilbk/nvim/blob/master/init.vim#L159) is how you can do that with Vim and [here](https://marketplace.visualstudio.com/items?itemName=mkhl.shfmt) is how for VScode.

jamilbk committed:
Add shell linter to CI

jamilbk committed:
remove typo

jamilbk committed:
Lint shell scripts

jamilbk committed:
Use ubuntu

jamilbk committed:
Fix migrate

jamilbk committed:
Install shfmt

jamilbk commented on swift/apple/lint.sh:
```diff
@@ -10,8 +10,8 @@ if [[ "$(uname -m)" == arm64 ]]; then
+    find . -name "*.swift" -not -path "./FirezoneNetworkExtension/Connlib/Generated/*" -exec xargs swift-format lint --strict \;
 else
-  echo "warning: swift-format not installed, install with `brew install swift-format`"
+    echo "warning: swift-format not installed, install with $(brew install swift-format)"
```
```suggestion
    echo "warning: swift-format not installed, install with 'brew install swift-format'"
```

jamilbk committed:
Update swift/apple/lint.sh

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
Add exit code

jamilbk committed:
Add to contributing guide

jamilbk commented on scripts/README.md:
@bmanifold Added just now here

ReactorScram commented on scripts/README.md:
```diff
@@ -0,0 +1,22 @@
+deployment of the Firezone product.
+
+## Developer Setup
+
```
```suggestion
`sudo apt-get install shellcheck`
Install shfmt from https://github.com/mvdan/sh/releases
```

Just put the exact commands up front, devs will read the rest if they need.

And I don't know how editor plugins work, I don't use any, so I'd offer the Bash equivalent:

```bash
shellcheck scripts/**/*.sh
shfmt -i 4 scripts/**/*.sh
```

Or maybe say "Make sure to set up pre-commit". I still need to do that on my system.

bmanifold replied:
@ReactorScram, if it's written in `go` then it's very unlikely that you'll need to install `go` to run it.  That's one thing that I really like about `go` is that it's so easy to build static binaries for a different architectures.

Looking at the releases page, it looks like they've already pre-built the binary for a lot of different systems:
https://github.com/mvdan/sh/releases

Also, it looks like they have a decent number of package manager integrations, so that could be another option.  Not sure how up to date the package managers are though.

ReactorScram replied:
Yeah turned out it's in apt, but WSL is only giving me 3.4.3. I'll change my suggestion to point at the releases page

jamilbk replied:
Thanks for the suggestions, updated

ReactorScram commented on CONTRIBUTING.md:
```diff
@@ -180,6 +181,11 @@ If you are interested in contributing to the Gateway, Relay, or client library,
+### Shell script Development
+
+If you are interested in contributing to any of our shell scripts, please read
+the detailed info found in the [Shell script README](scripts/README.md).
```
```suggestion
See [scripts/README](scripts/README.md).
```

ReactorScram commented on rust/README.md:
````diff
@@ -0,0 +1,19 @@
+rustup show
+```
+
+Then installed necessary targets depending on your platform and component.
````
```suggestion
Install rustup from https://rustup.rs/
Run `cargo --version` to make sure you have Rust installed.
```

Rustup installs the toolchain for the host platform by default.

ReactorScram replied:
Wait, i meant to make these part of a review. Sorry, Github.

jamilbk replied:
Updated

ReactorScram commented on rust/README.md:
```diff
@@ -0,0 +1,19 @@
+Firezone uses Rust for all data plane components in the product. This directory
+contains all of the low-level networking immplementations related to STUN/TURN,
+and the Rust-based Linux and Windows clients which are contained in this
+directory.
```
```suggestion
Firezone uses Rust for all data plane components. This directory
contains the Linux and Windows clients, and low-level networking implementations related to STUN/TURN.
```

Made it shorter.

ReactorScram reviewed (approved):
I had some suggestions on the README text. They're not blocking but every time I see a Github page I think "Just show me the commands. Just show the commands, I'll figure it out. Just put the install command right up top."

jamilbk committed:
set -euo pipefail

jamilbk committed:
Update CONTRIBUTING.md

Co-authored-by: Reactor Scram <ReactorScram@users.noreply.github.com>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
Update rust/README.md

Co-authored-by: Reactor Scram <ReactorScram@users.noreply.github.com>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
simplify README

jamilbk committed:
fix(ci): Ensure integration-tests allow for at least 30 seconds to establish a connection (#3676)

So the cause of the flaky tests is that they aren't waiting long enough
for a connection to be established. Both the test in #3666 and the
`iperf` tests have a timeout of 10 seconds.

Connections _should_ be established **very quickly** in CI. However, I
have a few guesses as to why they might not be, essentially causing us
to have to wait for a timeout to re-initiate a connection request:

- Packets arrive out of order or too quickly for the WireGuard state
machine to establish a handshake.
- Too many ICE candidates gathered (the gateway has 3 interfaces)


This PR:

- Refactors the iperf tests to be a little easier to maintain
- Ensures `integration-tests` run for at least 30 seconds before timing
out


In any case, we can debug / optimize this further after snownet is
merged, which might just solve the problem completely.

jamilbk committed:
rust readme

bmanifold committed:
feat(portal): Add sign-in success page for clients (#3659)

Why:

* On some clients, the web view that is opened to sign-in to Firezone is
left open and ends up getting stuck on the Sign In page with the
liveview loader on the top of the page also stuck and appearing as
though it is waiting for another response. This commit adds a sign-in
success page that is displayed upon successful sign-in and shows a
message to the user that lets them know they can close the window if
needed. If the client device is able to close the web view that was
opened, then the page will either very briefly be shown or will not be
visible at all due to how quickly the redirect happens.

Closes #3608 

<img width="625" alt="Screenshot 2024-02-15 at 4 30 57 PM"
src="https://github.com/firezone/firezone/assets/2646332/eb6a5df6-4a4c-4e54-b57c-5da239069ea9">

---------

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>
Co-authored-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
Merge branch 'main' into chore/ci-shell-lint

jamilbk committed:
Revert "feat(portal): Add sign-in success page for clients" (#3692)

Merged a bit too soon!

AndrewDryga commented on rust/docker-init.sh:
```diff
@@ -1,32 +1,32 @@
-  fi
+    if [ "${PUBLIC_IP4_ADDR}" = "" ]; then
+        public_ip4=$(curl "http://metadata.google.internal/computeMetadata/v1/instance/network-interfaces/0/access-configs/0/external-ip" -H "Metadata-Flavor: Google" -s)
+        export PUBLIC_IP4_ADDR="${public_ip4}"
```
@jamilbk do we use 4 spaces tab spacing for shell scripts? Elixir and Rust are using 2 spaces so it will lead to constant conflicts (editors will do one thing or another and CI will fail forcing people to go back and change spacing).

bmanifold replied:
@AndrewDryga, I know in Vim you can set your tab preferences on a per file type basis.  Could the same be done for VSCode? (I'm guessing yes, but I'm a VSCode n00b)

ReactorScram replied:
Rust is using 2 spaces? I thought Rust was using 4 spaces.

jamilbk replied:
Rust is 4 spaces -- I used 4 for shell scripts since that's the convention that they all were (except the ones in the swift/ and rust/ dir).

We can set to whatever. By default, `shfmt` uses 2 tabs I believe.

AndrewDryga replied:
Ah, did not know rust uses tabs. Let's use whatever (VSCode will handle that for me).

dependabot[bot] committed:
build(deps): Bump semver from 1.0.21 to 1.0.22 in /rust (#3685)

Bumps [semver](https://github.com/dtolnay/semver) from 1.0.21 to 1.0.22.
<details>
<summary>Release notes</summary>
<p><em>Sourced from <a
href="https://github.com/dtolnay/semver/releases">semver's
releases</a>.</em></p>
<blockquote>
<h2>1.0.22</h2>
<ul>
<li>Fix unused_imports warnings when compiled by rustc 1.78</li>
</ul>
</blockquote>
</details>
<details>
<summary>Commits</summary>
<ul>
<li><a
href="https://github.com/dtolnay/semver/commit/c8ad1bf6db35b1847abbe8ddc7c8fba0455e5fc3"><code>c8ad1bf</code></a>
Release 1.0.22</li>
<li><a
href="https://github.com/dtolnay/semver/commit/f76db8d7f2f542ac45aa2d7e3c06afa428fe8698"><code>f76db8d</code></a>
Resolve redundant import warning</li>
<li><a
href="https://github.com/dtolnay/semver/commit/f32b420f75d060cdc451bfac091318ed8bc0e41f"><code>f32b420</code></a>
Ignore incompatible_msrv clippy lint for conditionally compiled
code</li>
<li>See full diff in <a
href="https://github.com/dtolnay/semver/compare/1.0.21...1.0.22">compare
view</a></li>
</ul>
</details>
<br />


[![Dependabot compatibility
score](https://dependabot-badges.githubapp.com/badges/compatibility_score?dependency-name=semver&package-manager=cargo&previous-version=1.0.21&new-version=1.0.22)](https://docs.github.com/en/github/managing-security-vulnerabilities/about-dependabot-security-updates#about-compatibility-scores)

Dependabot will resolve any conflicts with this PR as long as you don't
alter it yourself. You can also trigger a rebase manually by commenting
`@dependabot rebase`.

[//]: # (dependabot-automerge-start)
[//]: # (dependabot-automerge-end)

---

<details>
<summary>Dependabot commands and options</summary>
<br />

You can trigger Dependabot actions by commenting on this PR:
- `@dependabot rebase` will rebase this PR
- `@dependabot recreate` will recreate this PR, overwriting any edits
that have been made to it
- `@dependabot merge` will merge this PR after your CI passes on it
- `@dependabot squash and merge` will squash and merge this PR after
your CI passes on it
- `@dependabot cancel merge` will cancel a previously requested merge
and block automerging
- `@dependabot reopen` will reopen this PR if it is closed
- `@dependabot close` will close this PR and stop Dependabot recreating
it. You can achieve the same result by closing it manually
- `@dependabot show <dependency name> ignore conditions` will show all
of the ignore conditions of the specified dependency
- `@dependabot ignore <dependency name> major version` will close this
group update PR and stop Dependabot creating any more for the specific
dependency's major version (unless you unignore this specific
dependency's major version or upgrade to it yourself)
- `@dependabot ignore <dependency name> minor version` will close this
group update PR and stop Dependabot creating any more for the specific
dependency's minor version (unless you unignore this specific
dependency's minor version or upgrade to it yourself)
- `@dependabot ignore <dependency name>` will close this group update PR
and stop Dependabot creating any more for the specific dependency
(unless you unignore this specific dependency or upgrade to it yourself)
- `@dependabot unignore <dependency name>` will remove all of the ignore
conditions of the specified dependency
- `@dependabot unignore <dependency name> <ignore condition>` will
remove the ignore condition of the specified dependency and ignore
conditions


</details>

Signed-off-by: dependabot[bot] <support@github.com>
Co-authored-by: dependabot[bot] <49699333+dependabot[bot]@users.noreply.github.com>

jamilbk committed:
chore(repo): Update issue template with support links (#3678)

Fixes #3627

---------

Signed-off-by: Jamil <jamilbk@users.noreply.github.com>
Co-authored-by: Andrew Dryga <andrew@dryga.com>

conectado committed:
fix(snownet): don't trigger busy loop upon refreshing allocation (#3693)

Currently, after the refresh timeout the gateway or client starts
looping forever, since after the refresh timeout is reached the
`poll_timeout` will always return `ALLOCATION_LIFETIME/2`(which would be
"now"), since `allocation_lifetime` is never updated.

To fix this, we need to check whether we currently have a refresh
request in flight before queuing a new one. Additionally, we need to
abort refreshing as soon as the lifetime of the allocation expires.

Related: #3617.
Related: #3631.

---------

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

thomaseizinger commented on rust/README.md:
```diff
@@ -0,0 +1,10 @@
+## Developer Setup
+
+- Install rustup from https://rustup.rs/
+- Run `cargo --version` to make sure you have Rust installed.
```
YMMV but not everybody uses rustup (I don't for example) so it would be more correct / helpful to say:

```suggestion
We target the last stable release of Rust using [`rust-toolchain.toml`](./rust-toolchain.toml).
If you are using `rustup`, that is automatically handled for you.
Otherwise, ensure you have the latest stable version of Rust installed.
```

ReactorScram replied:
Huh. Do you install Cargo with Nix or something?

ReactorScram replied:
I mentioned rustup caused I figured it's what novices use, and anyone who doesn't use rustup knows more than the README file anyway

thomaseizinger replied:
> Huh. Do you install Cargo with Nix or something?

Yeah I have an overlay where I can specify, which Rust version I want and it is set to latest stable. I haven't yet gotten around to setting up dev-shells properly but the idea would be that I can control per dev-shell which version I want.

thomaseizinger replied:
> I mentioned rustup caused I figured it's what novices use, and anyone who doesn't use rustup knows more than the README file anyway

Yeah that makes sense. I guess an expert can also find the rust-toolchain file by themselves :)

ReactorScram replied:
Looks like asdf can handle it too. I just started trying it a day or two ago after noticing it in the CONTRIBUTING.md doc

thomaseizinger reviewed (approved):
Big +1 for consistent formatting & linting as much as possible.

conectado committed:
refactor(connlib): replace `webrtc-rs` with `snownet` (#3391)

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

Resolves: #3377.

---------

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

jamilbk committed:
Update rust/README.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>
Signed-off-by: Jamil <jamilbk@users.noreply.github.com>

jamilbk committed:
Merge branch 'main' into chore/ci-shell-lint

jasonboukheir reviewed (approved):
👍 Good stuff. I appreciate the shell commands as a baseline :)

jamilbk committed:
Merge branch 'main' into chore/ci-shell-lint
